### PR TITLE
Migrate from github.com/golang/mock to go.uber.org/mock

### PR DESCRIPTION
**Pull request by Groxx**
`github.com/golang/mock` has been frozen for over a year and abandoned for even longer,
and is fully replaced by `go.uber.org/mock`.

So let's migrate.

In this PR, this is basically a no-op, but it forces some other upgrades (including Go).
For future code, this allows us to mock generic interfaces - the old library does not understand generics, and produces unusable code.

Most changes are trivial re-generation or import rewrites, both automated.
The remainders are:
- go.mods and related files, the makefile, and build/CI environment stuff
- `common/resource/resource_test_utils.go`
- `common/task/fifo_task_scheduler_test.go`
- `service/history/engine/testdata/engine_for_tests.go`
- `tools/common/schema/updatetask_test.go`

Which you can easily find by:
```sh
git diff master --stat | grep -v '2 +-' | grep -v mock
#                        ^ ignore import-only changes
#                                         ^ ignore mocks
```
and I've left comments on those `.go` files in the PR.

---

Steps to do this were:

1. `cd internal/tools`, `go get go.uber.org/mock/mockgen@latest`, and `go mod tidy` to upgrade the generator.
   - This didn't remove the dependency because lots of other libraries still use it too.
2. `go get go.uber.org/mock@latest` at the top level and DO NOT `go mod tidy` because it's unused, but it will be used by the next step.
   - Tidying now will simply remove it, as it didn't exist before, undoing this step.
3. Update the makefile to the new binary import path, and `make go-generate` to re-generate all mocks
   - If you don't `go get` the library first, this chokes if it re-parses anything it just generated, because it generated code that uses `go.uber.org/mock` but it doesn't have that dependency so it can't type-check it so it fails on its own output while producing its own output.  Fun!
   - `go.uber.org/mock` *could probably* streamline this by replacing `github.com/golang/mock` in the type checker with itself (`golang.org/x/tools/go/packages` has this ability), as it is almost identical.  But it's likely not worth the effort, and it likely cannot be made perfect.
4. `gofmt -w -r '"github.com/golang/mock/gomock" -> "go.uber.org/mock/gomock"' **/*.go` to rewrite all imports.
   - If you run this exact command in zsh, it'll also rewrite files in the idl submodule.  Just reset it.
5. `make build` will now fail in a few places, where `go.uber.org/mock` controllers are passed to the client library's mocks, which require `github.com/golang/mock` controllers.
   - Fix those by hand.
6. `make pr` to re-tidy now that the code's ready and refresh everything to make sure it's stable.
7. `make test` to validate, fix remaining issues 

Not too hard tbh, aside from the minor bootstrapping issues.
And thankfully we only construct the client mocks in a couple places.

Anyone with custom code can follow similar steps to migrate their test code / custom mocks / etc.
In a more ideal setup, where someone already has both dependencies, this will simplify to:
- switch the generator binary
- `go generate`
- `gofmt -r ...`
- `go mod tidy`

And likely no other steps, unless they also use other-library old-gomocks (like this repo does, for the client library).

---

The client library is still using `github.com/golang/mock` and unfortunately this can't
really be changed because it's a major breaking change.
Users can't even `replace old => new` because they get this error if *any other code or library*
uses `go.uber.org/mock`:
```
go: go.uber.org/mock@v0.5.0 used for two different module paths (github.com/golang/mock and go.uber.org/mock)
```
I don't really see why this would be blocked, as it seems rather important for any "compatible fork of X" migration process.
But Go has made the decision, so we're stuck with it.

So the only option to transparently upgrade is apparently to do what our internal monorepo
does, and dynamically rewrite ALL old-imports to new ones, for all code and all libraries.
(this works fine because they're essentially identical libraries).

This is all fairly awful.  But there's not much a library-user can do, aside from not
replicating the two core mistakes that got us here, when writing their own libraries:
- Don't make code generators that produce code that depends on the generator (or, ideally, anything outside stdlib).  Applications and libraries have wildly different versioning needs/desires:
  - Apps can and should update to latest frequently, because upgrades are useful and breaking changes can be fixed directly.
  - Libraries *must not* make breaking changes, because they're forcing breaking changes for all their users, and possibly conflicting versions on any other libraries someone else might use.
  - When these are combined, it's *very common* for a simple upgrade in the generator to lead to major transitive upgrades (often breaking) for library users, despite not being necessary.  This is very easy to miss while developing the tool itself.
- Libraries should not provide their own trivial mocks or convenience-interfaces.
  - Go has recommended ^ avoiding this from the very beginning.
  - It's normal advice for mocks and implicit interfaces in any language.
  - It's very commonly ignored.  And users complaining about it are correct, it's not great ergonomics in Go... but the alternative is eventually *much* worse, they need to be forced to do their part instead.
  - (Our workflow test suite is a clear exception, as it is not trivial.  But the RPC clients have no need to be interfaces, nor to have trivial generated mocks.)

So in here we just keep using the old mock controller for the old mocks, and upgrade everything
else to use the new ones.  It'd be nice to completely eliminate it, but it's not necessary at all.

### CODE REVIEW

File: common/resource/resource_test_utils.go [hunk 1]
```diff
@@ -23,11 +23,12 @@ package resource
 import (
 	"testing"
 
-	"github.com/golang/mock/gomock"
+	oldgomock "github.com/golang/mock/gomock" // client library cannot change from the old gomock
```
Comment (Groxx) [inline common/resource/resource_test_utils.go:26]: non-trivial change

File: common/task/fifo_task_scheduler_test.go [hunk 2]
```diff
@@ -83,7 +83,7 @@ func (s *fifoTaskSchedulerSuite) TestFIFO() {
 	tasks := []PriorityTask{}
 	var taskWG sync.WaitGroup
 
-	calls := []*gomock.Call{
+	calls := []any{
```
Comment (Groxx) [inline common/task/fifo_task_scheduler_test.go:86]: non-trivial change (`gomock.InOrder` below changed its type signature)

File: service/history/engine/testdata/engine_for_tests.go [hunk 3]
```diff
@@ -23,10 +23,11 @@ package testdata
 import (
 	"testing"
 
-	"github.com/golang/mock/gomock"
+	oldgomock "github.com/golang/mock/gomock" // client library cannot change from old gomock
```
Comment (Groxx) [inline service/history/engine/testdata/engine_for_tests.go:26]: non-trivial change

File: tools/common/schema/updatetask_test.go [hunk 4]
```diff
@@ -720,7 +720,7 @@ func (s *UpdateTaskTestSuite) Test_readManifest() {
 			subDir:        string([]byte{0xff, 0xfe, 0xfd, 0xfc}),
 			manifestFile:  manifestFileName,
 			invalidSubDir: true,
-			err:           errors.New("invalid name"),
+			err:           fs.ErrInvalid, // bad utf8 returns this as of 1.23: https://github.com/golang/go/commit/bf821f65cfd61dcc431922eea2cb97ce0825d60c
```
Comment (Groxx) [inline tools/common/schema/updatetask_test.go:723]: non-trivial change (go upgrade led to new behavior)

**Review by taylanisikdemir — APPROVED** [review]

**Review by davidporter-id-au — APPROVED** [review]
Excellent to read, ty for the effort in documenting the process

**Pull request merged**